File: problems/0019-remove-nth-node-from-end-of-list/solution.py
"""
Problem Number: 19
Problem Name: Remove Nth Node From End of List
Difficulty: Medium
Topics: Linked List, Two Pointers
Companies: Facebook, Amazon, Apple
"""

from typing import Optional, List

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """
        Approach: One Pass with Two Pointers
        Time Complexity: O(L)
        Space Complexity: O(1)
        """
        dummy = ListNode(0, head)
        first = dummy
        second = dummy
        
        # Move first n + 1 steps
        for _ in range(n + 1):
            first = first.next
            
        # Move until first reaches end
        while first:
            first = first.next
            second = second.next
            
        # Remove
        second.next = second.next.next
        
        return dummy.next

# Helper methods for testing
def create_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for val in arr:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

def list_to_arr(head):
    res = []
    while head:
        res.append(head.val)
        head = head.next
    return res

# Test cases
if __name__ == "__main__":
    solution = Solution()
    head = create_list([1,2,3,4,5])
    new_head = solution.removeNthFromEnd(head, 2)
    print(list_to_arr(new_head)) # [1,2,3,5]

File: problems/0019-remove-nth-node-from-end-of-list/solution.js
/**
 * @problemNumber: 19
 * @problemName: Remove Nth Node From End of List
 * @difficulty: Medium
 * @topics: Linked List, Two Pointers
 * @companies: Facebook, Amazon, Apple
 */

// Definition for singly-linked list.
function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

/**
 * Approach: One Pass with Two Pointers
 * Time Complexity: O(L)
 * Space Complexity: O(1)
 *
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
function removeNthFromEnd(head, n) {
    const dummy = new ListNode(0);
    dummy.next = head;

    let first = dummy;
    let second = dummy;

    // Move first pointer n + 1 steps ahead
    for (let i = 0; i <= n; i++) {
        first = first.next;
    }

    // Move both pointers until first reaches the end
    while (first !== null) {
        first = first.next;
        second = second.next;
    }

    // Skip the desired node
    second.next = second.next.next;

    return dummy.next;
}

// Helper to create list from array
function createList(arr) {
    let dummy = new ListNode(0);
    let curr = dummy;
    for (let val of arr) {
        curr.next = new ListNode(val);
        curr = curr.next;
    }
    return dummy.next;
}

// Helper to toArray
function toArray(head) {
    let res = [];
    while (head) {
        res.push(head.val);
        head = head.next;
    }
    return res;
}

// Test cases
let head = createList([1, 2, 3, 4, 5]);
console.log(toArray(removeNthFromEnd(head, 2))); // [1,2,3,5]

module.exports = { removeNthFromEnd };
